File: 139-word-break/word-break.cpp
class Solution {
public:
    int find(int i, vector<int> & dp, string s, vector<string>&words, int n){
        if (dp[i]==-1){
            for (string word:words){
                if (s.substr(i,word.size())==word){
                    if(find(i+word.size(),dp,s,words,n)){return dp[i]=1;}
                }
            }
            dp[i]=0;
        }
        return dp[i];
    }
    bool wordBreak(string s, vector<string>& words) {
        int n=s.size();
        vector<int> dp(n+1,-1);
        dp[n]=1;
        return bool(find(0,dp,s,words,n));
    }
};

File: 139-word-break/word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        wordDict=set(wordDict)
        n=len(s)
        dp=[-1]*(n+1)
        dp[0]=1
        maxlen = max(map(lambda x: len(x), wordDict))
        for i in range(n+1):
            for j in range(i,max(i-maxlen,0),-1):
                if dp[i]==-1:
                    if s[j-1:i] in wordDict and dp[j-1]==1:
                        dp[i]=1
            if dp[i]==-1:
                dp[i]=0
        return dp[n]==1
